[IV-23/24] Objetivo 5

diagmatrix opened the pull request:
* [x] ¿Se han establecido criterios previos de elección de la imagen base, así
como los criterios de búsqueda que han conducido a las imágenes que se están
considerando?
* [x] ¿Se ha tenido en cuenta que la imagen es para testear, y que por tanto
tiene una función específica, tal como testear cómo va nuestra aplicación en la
última versión publicada del lenguage?
* [x] ¿Se ha tenido en cuenta que la versión de la imagen base sea la última?
* [x] ¿Se han tenido en cuenta criterios reales para la elección, y no criterios
irrelevantes como cuanto tiempo se tarda en construirla? (Se va a construir una
vez, se va a usar cientos o miles de veces)
* [x] ¿Se han seguido las buenas prácticas en las órdenes que construyen
   la imagen? ¿Especialmente las relativas a usuarios no privilegiados y demás?
* [x] ¿La imagen incluye sólo y exclusivamente la infraestructura
   necesaria para pasar los tests? Por ejemplo, si se ha elegido la imagen
   "oficial" (lo que hay que hacer sólo en casos muy justificados) ¿se ha
   comprobado que lo que se instale sea sólo y exclusivamente lo necesario para
   nuestro proyecto?
_Por lo que he visto y entendido de la [imagen](https://hub.docker.com/layers/library/golang/1.22.0-alpine3.18/images/sha256-68fcc123b72e6f76ccee9680522217ba7f33a19b37f7d92ba1d8a10d483d5745?context=explore) sí, pero no elimino `go.mod` y `go.sum` porque supone muchos dolores de cabeza._
* [x] ¿Se han documentado y enlazado los commits a las imágenes que se
   han testeado?
_Sí, en [74ccf2e](https://github.com/diagmatrix/maybe-better-maybe-worse/commit/74ccf2efb06ca9801beac80a2f6074ec2d385de9) y [46ea6e8](https://github.com/diagmatrix/maybe-better-maybe-worse/pull/29/commits/46ea6e8f6077744358b7da311eff8e3b00311242)._
* [x] ¿Está establecido correctamente como `WORKDIR` el que se va a usar en los
   tests que se van a lanzar?
* [x] ¿Se usa algún directorio adecuado para copiar contenidos del repositorio
en vez del directorio de trabajo?
* [x] ¿Se han establecido claramente los criterios de búsqueda y de aceptación de
   la imagen base?
* [x] ¿Tienes claro cuales son las condiciones en las que tiene que actualizarse
la imagen en DockerHub y has configurado el workflow de acuerdo con ello? En
particular, ¿sabes qué condiciones tienen que darse para que se actualice y qué
condiciones *no* deben darse?
_Únicamente si se cambia la imagen en sí o se actualizan las dependencias del proyecto._
* [x] Si se usan GitHub actions para su publicación, ¿se han considerado
cuidadosamene los criterios (ramas, paths) para que se lancen?
* [x] ¿Se ha comprobado que el ENTRYPOINT del contenedor no escriba nada en el
directorio que monta, ya que no tiene derecho de escritura?
* [x] ¿Se ha abierto un shell en el contenedor una vez construida la imagen y
comprobado que los ficheros que se necesitan están todos incluidos, y que no hay
ficheros que no se necesiten?
_Sí, usando `sh`. No se ha encontrado ningún fichero del proyecto que no fuese necesario:_
```sh
/app/test # ls
/app/test # cd ..
/app # ls
go.mod        go.sum        taskfile.yml  test
/app # cd ..
/ # ls
app    bin    dev    etc    go     home   lib    media  mnt    opt    proc   root   run    sbin   srv    sys    tmp    usr    var
```

diagmatrix committed:
Selecciono imagen base #27

diagmatrix committed:
Añado Dockerfile  #28

diagmatrix committed:
Modifico Dockerfile #28

Elimino errores tipográficos

diagmatrix committed:
Añado imagen al README #28

diagmatrix committed:
Añado imagen al README #27

diagmatrix committed:
Merge branch 'objetivo-5' of github.com:diagmatrix/maybe-better-maybe-worse into objetivo-5

diagmatrix committed:
Añado fichero de automatización #30

diagmatrix committed:
Modifico Dockerfile #30

Modifico dockerfile para testear el workflow

diagmatrix committed:
Arreglo error #30

Arreglo un error de indentación

diagmatrix committed:
Arrelglo GitHub action #30

Faltaba trigger

diagmatrix committed:
Actualizo dependencias para testeo #30

Ejecuto "go mod tidy" para hacer trigger de la acción

diagmatrix committed:
Arreglo automatización #30

diagmatrix committed:
Arreglo indentación #30

diagmatrix committed:
Arreglo login #30

Elimino errata en los secretos

diagmatrix committed:
Modifico imagen para testeo #30

JJ commented on Dockerfile:
```diff
@@ -0,0 +1,27 @@
+# Base image for container
+FROM golang:alpine3.18
```
No debéis usar una versión específica, y menos si no es la última.

diagmatrix replied:
No entiendo muy bien, te refieres a que tendría que haber puesto `golang:alpine`?

JJ replied:
Nunca te voy a decir que pongas esto o lo otro, la decisión la tienes que tomar tú. Pero [si miras en los errores frecuentes](https://github.com/JJ/IV-/blob/master/errores/errores-objetivo-5.md) dice:

> De la misma forma, basarse sólo en el tamaño puede conducir a elegir imágenes cuya velocidad a la hora de ejecutar los tests no sea óptima.

En tu caso no has testeado la velocidad de los tests, pero el punto sigue siendo que elegir *la versión* de la imagen base en función del tamaño crea deuda técnica, porque al no ser la última versión es algo que hay que comprobar periódicamente a ver si hay errores de seguridad o de otro tipo (no es que tengan mucha importancia en una imagen de test, pero en todo caso puede tenerlo si luego se extiende a una de despliegue); en todo caso no es justificable usar una versión del SO base en vez de simplemente la última versión, que tendrá todo corregido y además se actualizará periódicamente cada vez que se cree la imagen.

diagmatrix replied:
Entiendo entonces que lo ideal si al final usase la oficial (no lo hago, arreglado aquí [3be7b2f](https://github.com/diagmatrix/maybe-better-maybe-worse/pull/29/commits/3be7b2f4d4289b103295648749adb7b8c9005824)), ¿tendría que usar `golang:latest` independientemente del SO base en el que esté construida? 

Por curiosidad, ¿ese SO puede cambiar? Las veces que he usado Docker anteriormente, he usado imágenes que sólo estaban en base a un SO, entonces no sé si eso puede pasar.

JJ commented on docs/imagen.md:
```diff
@@ -0,0 +1,50 @@
+
+| Criterio              | Ajuste en herramienta | 
+|------------------------|-----------------------|
+| *Tamaño* | La última versión de la imagen pesa 67.59 MB comprimida. |
```
No se debe usar una versión específica, porque genera deuda técnica.

JJ reviewed (changes requested):
Aparte de lo comentado, deberías explorar otras imágenes también, si es posible. Y siento no haber visto esto antes, se ve que me llegó la solicitud de revisión a spam y cuando lo miré por primera vez todavía no estaba enviado.

JJ commented:
>     * [ ]  ¿La imagen incluye sólo y exclusivamente la infraestructura
>       necesaria para pasar los tests? Por ejemplo, si se ha elegido la imagen
>       "oficial" (lo que hay que hacer sólo en casos muy justificados) ¿se ha
>       comprobado que lo que se instale sea sólo y exclusivamente lo necesario para
>       nuestro proyecto?
>       _Por lo que he visto y entendido de la [imagen](https://hub.docker.com/layers/library/golang/1.22.0-alpine3.18/images/sha256-68fcc123b72e6f76ccee9680522217ba7f33a19b37f7d92ba1d8a10d483d5745?context=explore) sí, pero puedo estar equivocado, por tanto no lo marco._

¿go.sum y go.mod son necesarios *una vez creada la imagen*?

> 
>     * [ ]  ¿Se han documentado y enlazado los commits a las imágenes que se
>       han testeado?
>       _No entiendo esta cuestión._

En general, no deberías decir el tamaño de la imagen base, sino el tamaño de la imagen construida, que va a tener un tamaño diferente de la base, y puede ser bastante más grande que el original. Pero para que sea creíble que lo has hecho, se deberían enlazar estos commits. Has hecho bien en no marcar si no estás seguro.
 
>     * [ ]  ¿Se usa algún directorio adecuado para copiar contenidos del repositorio
>       en vez del directorio de trabajo?
> 

Esto sería que si, ¿no?

JJ commented on Dockerfile:
```diff
@@ -0,0 +1,27 @@
+WORKDIR /app
+
+# Install task runner
+RUN go install github.com/go-task/task/v3/cmd/task@latest
```
¿Con qué privilegios estás instalando esto?

JJ commented on Dockerfile:
```diff
@@ -0,0 +1,27 @@
+WORKDIR /app/test
+
+# Set entrypoint for testing
+ENTRYPOINT ["task", "test"]
```
Este `task`, ¿qué taskfile está usando?

diagmatrix replied:
el que copia en la línea 18, ¿no?

JJ reviewed (changes requested):
Aparte de lo comentado, realmente me sentiría más cómodo si hubieras elegido `just`, como al principio, y otra imagen en vez de `bitnami`

JJ commented on go.mod:
```diff
@@ -2,10 +2,10 @@ module github.com/diagmatrix/maybe-better-maybe-worse
 require (
 	github.com/davecgh/go-spew v1.1.1 // indirect
 	github.com/pmezard/go-difflib v1.0.0 // indirect
-	github.com/stretchr/objx v0.5.0 // indirect
```
Esto ¿por qué ha desaparecido de aquí?

diagmatrix replied:
Lo comento más abajo, pero es el resultado de `go mod tidy`

JJ commented on docs/imagen.md:
```diff
@@ -0,0 +1,50 @@
+
+## Elección
+
+Por los criterios descritos, vamos a utilizar la imagen oficial de alpine, puesto que no tiene vulnerabilidades conocidas, es la más pequeña y es muy popular.
```
Hay muchas más imágenes que se podían considerar, incluso algunas aparecen en Docker Hub antes que la de `bitnami`. Tengan más o menos tamaño, hay que considerarlas.

JJ replied:
Como te he dicho arriba, conviene que pongas dónde la las resuelto, en qué commit, así se puede juzgar por parte del revisor si efectivamente se ha hecho. Me da la impresión de que has añadido Jumpserver, ¿crees que con añadir una más es suficiente? ¿O puede que haya otras imágenes, incluso de Alpine y Go, que no hayas incluido?

diagmatrix replied:
Arreglado aquí [3be7b2f](https://github.com/diagmatrix/maybe-better-maybe-worse/pull/29/commits/3be7b2f4d4289b103295648749adb7b8c9005824)

JJ commented on go.mod:
```diff
@@ -2,10 +2,10 @@ module github.com/diagmatrix/maybe-better-maybe-worse
 
 go 1.21.4
 
+require github.com/stretchr/testify v1.8.4
```
¿por qué ha cambiado esto de lugar? ¿Qué tiene que ver con el Dockerfile?

diagmatrix replied:
No sabía como hacer un test de la acción sin hacer un commit sobre uno de los archivos que la triggereaban, entonces ejecuté el comando `go mod tidy`, que por lo que entendí "optimiza" las dependencias a paquetes

JJ replied:
Podías haberle quitado comentarios al Dockerfile. O cambiado la versión.
En todo caso, el commit que cambia esto apunta a #30, que no explica el cambio. Siempre es conveniente que los cambios se expliquen adecuadamente, aunque en este caso es cierto que el mensaje de commit dice lo mismo que dices aquí.

JJ commented on docs/imagen.md:
```diff
@@ -0,0 +1,50 @@
+
+ - **Tamaño**: Buscaremos minimizar el tamaño de la imagen para optimizar el espacio.
+ - **Popularidad**: Buscaremos usar las imágenes más populares posibles, basándonos en los pulls de Docker Hub.
+ - **Mantenimiento**: Se descartarán las imágenes obsoletas, eligiendo aquellas que reciban actualizaciones frecuentes.
```
¿La seguridad no te parece importante? Puede que haya razones de seguridad importantes para rechazar alguna de las imágenes que se incluyen ahí abajo.

JJ replied:
No deberías resolver ninguna conversación sin cambio o comentario indicando donde se ha resuelto (parece que abajo, pero la costumbre como has hecho en otros lugares es poner el commit en el que se ha hecho)

JJ replied:
Igual que te he comentado antes, no deberías resolver sin indicar dónde se ha hecho.

JJ reviewed (changes requested):
Alguna cosa adicional que deberías tener en cuenta.

diagmatrix commented:
> Aparte de lo comentado, realmente me sentiría más cómodo si hubieras elegido `just`, como al principio, y otra imagen en vez de `bitnami`

Salvo que haya cometido algún error escribiendo, he elegido la imagen oficial de Go con `alpine`

diagmatrix committed:
Añado posibles alternativas #27

diagmatrix committed:
Compruebo tamaños imágenes #27

diagmatrix committed:
Modifico elección de gestor de tareas #16

Cambio de task a just

diagmatrix committed:
Actualizo las órdenes en el README #16

Al modificar de task a just

diagmatrix committed:
Modifico imagen #28

Cambio de task a just y elimino archivos innecesarios.

diagmatrix committed:
Añado enlace del contenedor generado  #30

diagmatrix commented:
@JJ listo para revisión

JJ commented on Dockerfile:
```diff
@@ -0,0 +1,30 @@
+    version="1.1"
+
+# Create user without privileges for the app
+RUN adduser --disabled-password -u 1001 ppf-tests
```
¿Cómo justificas deshabilitar el password? ¿Y el UID?

diagmatrix replied:
Cambio el UID para que no use el mismo espacio de nombres de usuarios que el demonio de Docker, evitando así alguna de las posibles vulnerabilidades que puedan hacer uso del `host` desde el contenedor.

El deshabilitar el password es para que no pueda hacerse un login interactivo, haciendo así que se ejecute el contenedor sin privilegios.

Me he dado cuenta de que en ningún momento cambio a ese usuario, así que lo voy a arreglar. ¿Tendría que haber escrito eso en un issue?

JJ replied:
> Cambio el UID para que no use el mismo espacio de nombres de usuarios que el demonio de Docker, evitando así alguna de las posibles vulnerabilidades que puedan hacer uso del `host` desde el contenedor.
> 

El usuario se puede cambiar al ejecutar el contenedor, y de hecho se hace. El usuario normal usa el UID 1000, ¿eso está en otro "espacio de nombres?" ¿Qué es un "espacio de nombres"? ¿En qué issue está expresado ese problema y qué HU corresponde? ¿Qué requisitos de seguridad tiene un contenedor *para ejecutar tests*?

> El deshabilitar el password es para que no pueda hacerse un login interactivo, haciendo así que se ejecute el contenedor sin privilegios.

Mismas preguntas que arriba. ¿Login interactivo en un contendor de pruebas que va a ejecutar tests? 
> 
> Me he dado cuenta de que en ningún momento cambio a ese usuario, así que lo voy a arreglar. ¿Tendría que haber escrito eso en un issue?

Si hubiera algún tipo de HU que lo requiriera, o te hubieras dado cuenta de algún error en la construcción, para eso están los issues.

diagmatrix replied:
Arreglado aquí [3be7b2f](https://github.com/diagmatrix/maybe-better-maybe-worse/pull/29/commits/3be7b2f4d4289b103295648749adb7b8c9005824)

JJ commented on Dockerfile:
```diff
@@ -0,0 +1,30 @@
+# Base image for container
+FROM golang:alpine
```
De [los errores frecuentes](https://github.com/JJ/IV-/blob/master/errores/errores-objetivo-5.md):

> El principal problema es limitarse a las imágenes oficiales; muchas de estas están creadas directamente por Docker y no tienen que ser demasiado óptimas; puede que lo sean en el tamaño, pero poco más. Además, una vez que se limita uno a las imágenes oficiales, no hay mucho margen de elección, porque la más pequeña suele ser la basada en Alpine.

Lo que se busca en esta metodología docente es que el estudiante se preocupe de considerar diferentes opciones para una decisión y, por supuesto, que la justifique. En Docker Hub hay varias docenas de imágenes diferentes, que no has considerado en la justificación más abajo.

diagmatrix replied:
Respondido aquí https://github.com/diagmatrix/maybe-better-maybe-worse/pull/29#discussion_r1486637686

JJ replied:
Te agradezco la respuesta, pero también te agradecería que leyeras los errores frecuentes que te he comentado, y cambiaras tu decisión.

diagmatrix replied:
Resuelto aquí [3be7b2f](https://github.com/diagmatrix/maybe-better-maybe-worse/pull/29/commits/3be7b2f4d4289b103295648749adb7b8c9005824)

JJ commented on Dockerfile:
```diff
@@ -0,0 +1,30 @@
+# Base image for container
```
Adicionalmente, es conveniente que sepáis realmente qué estáis usando. [Esta es la plantilla para las imágenes oficiales](https://github.com/docker-library/golang/blob/master/Dockerfile-linux.template) ¿Necesitas todo lo que hay ahí?  Por otro lado, ¿has visto si hay instalado un gestor de tareas que puedas usar? ¿O si hay un usuario ya creado?

diagmatrix replied:
Viendo las layers [aquí](https://hub.docker.com/layers/library/golang/alpine/images/sha256-3325c5593767d8f1fd580e224707ca5e847a1679470a027aaa3c71711ce16613?context=explore), la 0 y la 2 instalan paquetes de alpine, entre ellos `apk`, el gestor de paquetes por defecto de la distribución y en la 7 se instala go, con el test runner que uso `go test` y el gestor de dependencias `go mod`. En cuanto a creación de usuarios, no parece

JJ replied:
Puedes ver las layers ahí como dices, que están bastante incompletas, o puedes mirar el Dockerfile que te he enlazado y ver que, por ejemplo, se instala make. 

https://github.com/docker-library/golang/blob/dae3405a325073e8ad7c8c378ebdf2540d8565c4/Dockerfile-linux.template#L153

Aparte de los compiladores de c y c++. ¿Necesitas eso para tus tests?

Por otra parte ¿realmente crees que rebatir los comentarios que hago en vez de actuar o tenerlos en cuenta contribuye a que yo entienda si has superado o no los diferentes objetivos de aprendizaje aquí?

diagmatrix replied:
Resuelto aquí [3be7b2f](https://github.com/diagmatrix/maybe-better-maybe-worse/pull/29/commits/3be7b2f4d4289b103295648749adb7b8c9005824)

En cuanto a rebatir, no era mi objetivo con el comentario, aunque re-leyendo es un poco pasivo-agresivo. Perdona

JJ commented on Dockerfile:
```diff
@@ -0,0 +1,30 @@
+COPY go.mod go.sum justfile ./
+
+# Download the dependencies
+RUN go mod download && go mod verify
```
¿Con qué privilegios estás haciendo esto?

diagmatrix replied:
Con los de root :fearful:, arreglado en [23cd058](https://github.com/diagmatrix/maybe-better-maybe-worse/pull/29/commits/23cd05861b79889c3be5615b35a54672828d94a9) y comentado aquí https://github.com/diagmatrix/maybe-better-maybe-worse/pull/29#discussion_r1486571986

JJ commented on docs/imagen.md:
```diff
@@ -0,0 +1,76 @@
+
+| Criterio              | Ajuste en herramienta | 
+|------------------------|-----------------------|
+| *Popularidad* | La imagen tiene más de medio millón de pulls en Docker Hub. |
```
Está bien que crees las tablas, pero has indicado arriba como uno de los criterios el tamaño y luego no lo has indicado en estas tablas.

JJ replied:
.... y deberías resolver también los comentarios indicando en qué commit se ha hecho.

diagmatrix replied:
Resuelto aquí [14439b5](https://github.com/diagmatrix/maybe-better-maybe-worse/pull/29/commits/14439b519fb280914125192b174fe2c249737bf2)

JJ commented on docs/imagen.md:
```diff
@@ -0,0 +1,76 @@
+| Criterio              | Ajuste en herramienta | 
+|------------------------|-----------------------|
+| *Mantenimiento* | La última imagen es de hace 5h (visitado el 7 de febrero de 2024). |
+| *Seguridad* | La última versión de esta imagen contiene más de 80 vulnerabilidades, siendo 2 de ellas severas (visitado el 11 de febrero de 2024). |
```
OK que esto es determinante como para no elegirla más allá del tamaño que tenga, pero tú mismo has indicado el tamaño como uno de los criterios. Si al final no lo vas a considerar, simplemente indica al principio que no se van a listar imágenes que tengan vulnerabilidades, mencionándolas si quieres. Pero te recuerdo que esto es una documentación de la decisión tomada, y por tanto debes ser lo más preciso posible justificando esa decisión.

JJ commented on docs/imagen.md:
```diff
@@ -0,0 +1,76 @@
+
+## Tamaño
+
+Para comprobar el tamaño de las imágenes, se ha ejecutado `docker pull <nombre_imagen>` sobre cada
```
Ah, OK que lo has puesto aquí. Habría sido más sencillo en otro lugar, pero en todo caso, ¿lo importante no será el tamaño una vez instalado `just` y el resto de los cambios en la imagen?
Por otro lado, si el tamaño es el principal criterio, ¿has mirado cuantas imágenes oficiales o de vendors oficiales de Alpine y Go hay en Docker Hub y qué tamaño tienen?
Lo que se busca en este objetivo no es que hagáis una imagen específica para un problema específico; se busca que aprendáis cómo se hace este tipo de cosas. Una de la mejor forma de hacerlo es buscar diferentes imágenes, ver como están construidas, y la vayas a elegir o no puedes ver diferentes patrones que te pueden ser interesantes para aplicarlos en tu trabajo.

diagmatrix replied:
Arreglado aquí [14439b5](https://github.com/diagmatrix/maybe-better-maybe-worse/pull/29/commits/14439b519fb280914125192b174fe2c249737bf2) (explico que se va a discutir en otro lugar)

JJ commented on docs/imagen.md:
```diff
@@ -0,0 +1,76 @@
+REPOSITORY          TAG        IMAGE ID       CREATED        SIZE
+bitnami/golang      latest     a53c682817d0   8 hours ago    611MB
+golang              alpine     a2742f74d90f   5 days ago     230MB
+golang              bullseye   3af37afcedb7   5 days ago     755MB
```
Con just y Go no es que tenga mayor importancia, porque en mi máquina tarda 2 milisegundos en hacer los tests, que va a ser una milésima parte de lo que tarda el flujo CI, pero en general, puede ser que las herramientas basadas en libc sean más rápidas que las basadas en `musl` como Alpine (aunque si quieres ver si la toolchain de Go está basada en una o en otra tendrás que mirar cómo están construidas las imágenes de Alpine y Go)

diagmatrix replied:
He hecho pruebas con la imagen de archlinux, se explica aquí: [14439b5](https://github.com/diagmatrix/maybe-better-maybe-worse/pull/29/commits/14439b519fb280914125192b174fe2c249737bf2)

JJ reviewed (changes requested):
Diría que ha mejorado en bastantes aspectos, pero es bastante importante que tengáis en cuenta una gama más amplia de imágenes base, y eso sin empezar a considerar simplemente imágenes base de sistemas operativos sobre las que se descarga exclusivamente lo que se necesita, la toolchain de go en este caso (un método que ayuda a alcanzar los objetivos de aprendizaje con más rapidez, ya que tienes que tener en cuenta todo lo necesario para el lenguaje *y* proyecto, en vez de tomar una imagen que lo tiene ya todo hecho)

diagmatrix committed:
Pruebo imagen archlinux #27, #28

Creo la imagen a partir de archlinux, documentando el Dockerfile del contenedor probado.

diagmatrix committed:
Modifico Dockerfile #28

Me aseguro de que el contenedor ejecuta los tests con un usuario sin privilegios, a cambio de no eliminar go.mod y go.sum

diagmatrix committed:
Mejoro la justificación de la elección #27

No cambio de decisión, pero describo las pruebas con otras imágenes

diagmatrix committed:
Documento decisiones de la imagen #28

diagmatrix commented:
@JJ listo para revisión. Documento en el último commit algunas de las cosas que me has comentado, como los permisos en el Dockerfile y esas cosas ([aquí](https://github.com/diagmatrix/maybe-better-maybe-worse/pull/29#discussion_r1485840410), por ejemplo)

JJ commented on Dockerfile:
```diff
@@ -0,0 +1,29 @@
+RUN apk add just
+
+# Create user without privileges for the app
+RUN adduser --disabled-password -u 1001 ppf-tests
```
Sigue sin haber justificación para que se cree con estas características. ¿Por qué 1001 y no 1100? ¿Por qué no el uid por omisión que es 1000? ¿Por qué se deshabilita el password, si nunca se va a hacer login interactivo?

diagmatrix replied:
Es cierto que no es necesario para este contenedor, lo cambio. De hecho, el uid 1001 estaba impidiendo ver que se escribían cosas en `app/test` sin yo saberlo (El caché de Go al lanzar los tests)

diagmatrix replied:
Solucionado aquí [3be7b2f](https://github.com/diagmatrix/maybe-better-maybe-worse/pull/29/commits/3be7b2f4d4289b103295648749adb7b8c9005824) y [edc43c7](https://github.com/diagmatrix/maybe-better-maybe-worse/pull/29/commits/edc43c759dec8ccf5906c3274adf1a2941f7a1af)

JJ commented on docs/archlinux_dockerfile.md:
````diff
@@ -0,0 +1,36 @@
+ENTRYPOINT ["just", "test"]
+```
+
+No elimino `go.mod` y `go.sum` porque no ocupan mucho espacio y era un rompecabezas intentar hacerlo 
````
¿Qué tipo de rompecabezas?

diagmatrix replied:
De permisos con el usuario creado, porque se cambia a él y no tiene permisos de escritura sobre `/app`

JJ replied:
¿Y qué problema hay con darle esos permisos?

diagmatrix replied:
Solucionado aquí [3be7b2f](https://github.com/diagmatrix/maybe-better-maybe-worse/pull/29/commits/3be7b2f4d4289b103295648749adb7b8c9005824) y [edc43c7](https://github.com/diagmatrix/maybe-better-maybe-worse/pull/29/commits/edc43c759dec8ccf5906c3274adf1a2941f7a1af)

JJ commented on Dockerfile:
```diff
@@ -0,0 +1,29 @@
+WORKDIR /app/test
+
+# Set entrypoint for testing
+ENTRYPOINT ["just", "test"]
```
¿Qué `justfile` usa esta orden?

diagmatrix replied:
Creo que entiendo lo que quieres decir, lo miro y veo si hace falta añadir el `justfile` al contenedor

diagmatrix replied:
Solucionado aquí [3be7b2f](https://github.com/diagmatrix/maybe-better-maybe-worse/pull/29/commits/3be7b2f4d4289b103295648749adb7b8c9005824) y [edc43c7](https://github.com/diagmatrix/maybe-better-maybe-worse/pull/29/commits/edc43c759dec8ccf5906c3274adf1a2941f7a1af)

JJ commented on docs/imagen.md:
````diff
@@ -0,0 +1,188 @@
+```Bash
+diagmatrix@gibbon ~ $ docker images
+REPOSITORY                            TAG         IMAGE ID       CREATED              SIZE
+diagmatrix/maybe-better-maybe-worse   archlinux   64b91dbeea80   About a minute ago   1.14GB
````
Comparar archlinux con alpine va a ser un poco injusto. Sólo cargar 1.14 Gb frente a 236 va a implicar una diferencia. ¿De veras que no has buscado simplemente `golang` y `alpine` para ver qué hay?

diagmatrix replied:
Creo que no he entendido tus comentarios anteriores respecto a esto. Voy a probar con `alpine` y ver que tal.

diagmatrix replied:
He visto también `alpinelinux/golang`, pero no tiene nada de información... La he descartado

diagmatrix replied:
Solucionado aquí [3be7b2f](https://github.com/diagmatrix/maybe-better-maybe-worse/pull/29/commits/3be7b2f4d4289b103295648749adb7b8c9005824)

JJ commented on docs/imagen.md:
````diff
@@ -0,0 +1,188 @@
+sys	0m0,010s
+```
+
+Podemos comprobar que en tiempo de ejecución casi idénticos.
````
Como te he dicho en mi comentario, es complicado que en algo que tarda 3 milisegundos se encuentre una diferencia.

diagmatrix replied:
¿Elimino entonces esta comprobación?

JJ replied:
Tú verás si te hace falta. Lo que yo buscaba con el comentario no es que hicieras la comprobación, sino que entendieras la diferencia entre los contenedores basados en Alpine y los basados en otras distros, y si realmente esa diferencia existe en el caso de Go. Alpine usa `musl' mientras que otros usan `libc`. Pero... tendrás que mirar como se construye el contenedor para ver si go usa realmente `musl` o realmente usa `libc` (que se puede instalar en Alpine).
Aparte de la biblioteca básica, puede haber otros factores que influyan en la velocidad relacionados por ejemplo con el scheduler, como se manejan las hebras y demás. No creo que ninguno tenga influencia en este caso.
TL;DR: Aparte que con algo que tarda 3 milisegundos es difícil que haya alguna diferencia entre diferentes sistemas (la teoría de ISE debería haberos explicado algo durante el benchmarking) lo que realmente perseguía yo con mi comentario es que entendáis las diferencias que hay entre diferentes imágenes base, aparte del tamaño.

diagmatrix replied:
Mis disculpas. Creo que no tiene sentido ponerlo. Al parecer, para compilar, Go utiliza preferiblemente la versión de `libc` que tenga instalada el SO (`musl` o `gnu`).

diagmatrix replied:
Solucionado aquí [3be7b2f](https://github.com/diagmatrix/maybe-better-maybe-worse/pull/29/commits/3be7b2f4d4289b103295648749adb7b8c9005824)

JJ reviewed (changes requested):
El principal problema sigue siendo la justificación de la imagen. Realmente sería conveniente que cuando se os dice que reconsideréis la decisión, *lo hagáis*. Podéis hacerlo de la forma que se os dice (buscando imágenes con las mismas características, por ejemplo) o de cualquier otra forma, pero de veras que deberíais intentar seguir los comentarios porque puede que aprendáis algo por el camino.
Por otro lado, la imagen tiene aspectos sin ningún tipo de justificación (password deshabilitado, por ejemplo, o un UID específico para el usuario) que "solucionan" un problema inexistente o no diagnosticado, y no soluciona problemas que sí existen como que haya en la misma ficheros que no son necesarios para la ejecución de los tests o ni siquiera necesarios para la construcción de la misma.
Por favor, trata de resolver esto a la mayor celeridad, es decir esta misma mañana, porque la presentación es mañana.

diagmatrix commented:
> Por favor, trata de resolver esto a la mayor celeridad, es decir esta misma mañana, porque la presentación es mañana.

Estoy en mi puesto de trabajo, me pondré en cuanto llegue después de almorzar

diagmatrix committed:
Modifico elección #27 #28

Discuto un nuevo criterio y mejoro la justificación del Dockerfile

diagmatrix committed:
Modifico imagen #28

Cambio imagen base a alpine

diagmatrix commented:
@JJ Listo para reivisión. Una cosa que me he dado cuenta es que se ha colado los cambios del Dockerfile en este commit [3be7b2f](https://github.com/diagmatrix/maybe-better-maybe-worse/pull/29/commits/3be7b2f4d4289b103295648749adb7b8c9005824) en lugar de en este [edc43c7](https://github.com/diagmatrix/maybe-better-maybe-worse/pull/29/commits/edc43c759dec8ccf5906c3274adf1a2941f7a1af). Mis disculpas.

TL;DR de los cambios es que he añadido como criterio la inspección de los paquetes que se instalan y he cambiado la imagen de `golang/alpine` a `alpine`, cambiando las cosas del Dockerfile que estaban mal

JJ reviewed (approved):
:+1: 
¿Cómo hacemos para la presentación?
